Fail build after several unsuccessfull retries

soltysh opened the pull request:
@smarterclayton after talking to @bparees we've decided we cannot loose the requirement for output image, because we use that information for [pushing](https://github.com/openshift/origin/blob/master/pkg/build/controller/controller.go#L96-L100). Without it we'd be guessing what user actually meant. In the end we've decided to go back to initial idea where we will fail a build upon X retires (where X is 60 currently, which is ~1 minute).

bparees commented:
lgtm.

smarterclayton commented:
Do not agree.  Let's discuss more.

----- Original Message -----
> lgtm.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2220#issuecomment-101747906

bparees commented:
@smarterclayton do you not agree we should actually give up on retrying the build, or do you not agree that a bad output image target should result in a failed build if the output target doesn't show up after some reasonable period of time?

bparees commented:
@soltysh I spoke with @smarterclayton and the conclusion we reached is that the retry period needs to be longer and time based instead of retry count based.  Can you update the retry checking logic to be timestamp based instead of count based?  and fail the build after 30 minutes from the first retry?

mfojtik commented:
@bparees I think 30 minutes is very short time, isn't it?

bparees commented:
@mfojtik not for a build that is failing because the configuration is broken/invalid.  if you didn't fix the config in 30 minutes, you're not paying attention and aren't going to fix it.  The real use case here is for asynchronous creation of things where the imagestream or secret or something else might get created a few seconds after the buildconfig that's being instantiated.

smarterclayton commented:
And it might take several minutes to import an image due to downtime.

> On May 15, 2015, at 10:14 AM, Ben Parees <notifications@github.com> wrote:
> 
> @mfojtik not for a build that is failing because the configuration is broken/invalid.  if you didn't fix the config in 30 minutes, you're not paying attention and aren't going to fix it. The real use case here is for asynchronous creation of things where the imagestream or secret or something else might get created a few seconds after the buildconfig that's being instantiated.
> 
> —
> Reply to this email directly or view it on GitHub.
>

soltysh commented:
I've rebased the code and added the 30 minutes timeout for build retries. I'm currently working on adding the events informations as per discussion in #2269.

soltysh commented:
Per @derekwaynecarr's suggestion I've just added showing build pod events in a build describe command. @bparees @smarterclayton  PTAL

[test]

smarterclayton commented on pkg/build/controller/factory/factory.go:
```diff
@@ -26,11 +26,29 @@ import (
+		build.Status = buildapi.BuildStatusFailed
+		build.Message = err.Error()
+		dummy := kutil.Now()
+		build.CompletionTimestamp = &dummy
```
Better name for this

smarterclayton commented on pkg/build/controller/factory/factory.go:
```diff
@@ -26,11 +26,29 @@ import (
+func limitedLogAndRetry(buildupdater buildclient.BuildUpdater) controller.RetryFunc {
+	return func(obj interface{}, err error, retries controller.Retry) bool {
+		kutil.HandleError(err)
+		if time.Since(retries.StartTimestamp.Time) < maxTimeout {
```
maxTimeout should be an argument to limitedLogAndRetry

smarterclayton commented on pkg/build/controller/factory/factory.go:
```diff
@@ -26,11 +26,29 @@ import (
+		dummy := kutil.Now()
+		build.CompletionTimestamp = &dummy
+		glog.V(3).Infof("Giving up retrying build %s/%s: %v", build.Namespace, build.Name, err)
+		if err := buildupdater.Update(build.Namespace, build); err != nil {
```
If you get a not found you shouldn't retry.

smarterclayton commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -108,12 +108,24 @@ func (d *BuildDescriber) Describe(namespace, name string) (string, error) {
+		events = &kapi.EventList{}
+	}
+	// get also pod events and merge it all into one list for describe
+	pod, _ := d.kubeClient.Pods(namespace).Get(buildutil.GetBuildPodName(build))
```
You can't eat the error here - that means you'll be passing nil to search.

smarterclayton replied:
Should be `if pod, err := ...; err == nil { add events }`

smarterclayton commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -135,7 +147,7 @@ func (d *BuildDescriber) Describe(namespace, name string) (string, error) {
 func describeBuildDuration(build *buildapi.Build) string {
 	t := util.Now().Rfc3339Copy()
-	if build.StartTimestamp == nil && build.Status == buildapi.BuildStatusCancelled {
+	if build.StartTimestamp == nil && (build.Status == buildapi.BuildStatusCancelled || build.Status == buildapi.BuildStatusFailed) {
```
What about BuildStatusError?

soltysh commented:
@smarterclayton addressed your comments.

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -151,12 +179,12 @@ func (factory *ImageChangeControllerFactory) Create() controller.RunnableControl
 					return false
 				}
-				if count > 60 {
+				if retries.Count > maxRetries {
```
please make this consistent with the implementation on line 139 above.

soltysh replied:
Done.

bparees commented on pkg/controller/controller_test.go:
```diff
@@ -80,13 +80,13 @@ func TestQueueRetryManager_retries(t *testing.T) {
-		retryFunc: func(obj interface{}, err error, count int) bool {
-			if count > 4 {
+		retryFunc: func(obj interface{}, err error, r Retry) bool {
+			if r.Count > 4 {
```
no tests for the new timestamp field?

soltysh replied:
I'll just add checking for non-zero value on this field, as other tests will unnecessarily increase the times of the test.

bparees commented on pkg/build/controller/factory/factory_test.go:
```diff
@@ -0,0 +1,44 @@
+	err := errors.New("funky error")
+
+	now := kutil.Now()
+	retry := controller.Retry{0, kutil.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute()-31, now.Second(), now.Nanosecond(), now.Location())}
```
needs a positive test that ensures young enough items get retried.

soltysh replied:
Done.

bparees commented:
one minor nit and a couple suggestions around testing, otherwise looks good!

soltysh commented:
@bparees addressed your comments. @smarterclayton waiting for your sing-off.

smarterclayton commented:
This is ba-ba-ba-brokENnnnnn

soltysh commented:
:cry: was working this morning, will fix it right away.

soltysh commented:
Aha, the problem is with recent changes to master (`pkg/deploy/controller/deploymentcancellation` package, which is missing in my PR). I need to rebase my PR to update necessary bits.

soltysh commented:
Fixed :smiley:

smarterclayton commented:
Lalalalaallalaaaaaaa....ooks good to meeEEEEeee

mfojtik commented:
@smarterclayton lol

soltysh commented:
[merge]
